tests/storage/test_base.py

```diff
@@ -209,3 +209,63 @@ def test_delete_one(self) -> Generator["defer.Deferred[object]", object, None]:
         self.mock_txn.execute.assert_called_with(
             "DELETE FROM tablename WHERE keycol = ?", ["Go away"]
         )
+
+    @defer.inlineCallbacks
+    def test_upsert(self) -> Generator["defer.Deferred[object]", object, None]:
+        self.mock_txn.rowcount = 1
+
+        result = yield defer.ensureDeferred(
+            self.datastore.db_pool.simple_upsert(
+                table="tablename",
+                keyvalues={"columnname": "oldvalue"},
+                values={"othercol": "newvalue"},
+            )
+        )
+
+        self.mock_txn.execute.assert_called_with(
+            "INSERT INTO tablename (columnname, othercol) VALUES (?, ?) ON CONFLICT (columnname)  DO UPDATE SET othercol=EXCLUDED.othercol",
```

Comment:
Aside: where does the extra space between `(columnname)` and `DO UPDATE` come from? (I guess this is just newlines getting converted to a regular space?)

Reply:
> I guess this is just newlines getting converted to a regular space?)

It is there in-case there's a where clause.